lib/web_ui/lib/ui_web/src/ui_web/initialization.dart

```diff
@@ -55,3 +55,22 @@ Future<void> bootstrapEngine({
     loader.didCreateEngineInitializer(bootstrap.prepareEngineInitializer());
   }
 }
+
+/// Loads manifest fonts.
+///
+/// By default manifest fonts are loaded automatically as part of engine
+/// initialization. However, that may delay the invocation of the app's `main`
+/// function. If the app needs to start doing useful work before the fonts are
+/// fully loaded, such as render UI without text (e.g. a loading indicator), or
+/// perform non-UI work (e.g. make early HTTP calls), then the configuration
+/// option `loadManifestFontsBeforeAppMain` can be set to `false`. This will
+/// make the engine skip font loading. Instead, the app can call this function
+/// directly when appropriate.
+///
+/// Beware that attempting to render text before fonts are loaded may lead to
+/// improper text rendering. It is expected that the developer has enough
+/// control over the app's code to ensure the correct loading sequence.
+Future<void> loadManifestFonts() async {
```

Comment:
Can we just call this method before engine initialization and then plumb the future through to runApp and await it there? To me, this seems like it'd be optimal for the vast majority of apps and avoid them adding more complexity in main.dart? Could we even change the default to do the above? I assume there would be minor consequences e.g. if people do some kind of custom prerendering things before runApp, they could break.

Reply:
This method includes fetching the fonts _and_ loading them into CanvasKit (or any renderer in general). So we need at least `initializeEngineServices` to complete, and that's a big chunk of engine initialization. I imagine, since this API is already for advanced users, those users will appreciate having more control despite a little more complexity. In particular, some users may want to skip manifest fonts entirely and rely purely on fonts loaded via `FontLoader`.

Reply:
IIUC `customer: quake` does want to render UI before fonts are downloaded (I'll double-check with them), so we may already have a use-case for not blocking `runApp`.

More generally, I feel like for the use-case of fine-tuning your app start-up sequence this kind of logic should not be a hidden implementation detail, but be part of an explicit API. I think there's more complexity in hiding it than exposing it, because instead of seeing a function invocation in your code and deducing its effect on that line (and also deciding for yourself where to call this function), you have to reason about how the font loading process changes behind-the-scenes inside the engine and in parallel with your logic.

Reply:
Re advanced users: Yeah, this makes sense... I take the point that there are advanced use cases that require full manual control, and so just disabling automatic font loading completely is best for them. So this CL seems ideal to unblock that case.

I guess what I'm wondering is whether there's an opportunity to boost the performance of every app that needs to do some asynchronous work in main() very cheaply by pushing people towards a default of fonts not blocking main() execution as Mouad was saying. Remember, John's font loading optimization approach (CL/560854521) improved overall loading time by ~9% overall for Earth for slower connections. What if you could give all new apps an improvement of this scale with a relatively small change to the contract of main() vs runApp() that would probably not affect most apps at all? Surely this is a cheaper optimization than trying to shave another 10% off the binary size of canvaskit etc?

I actually didn't even realise there was a distinction between loading fonts at the start or end of engine initialization, though if it's possible to kick off the RPC as early as possible and then load the fonts into the engine when the engine is ready, that would be great.

Reply:
Sorry I wrote my comment before I'd read yours, Yegor - that answers most of my concerns. Fair point that often having access to the actual futures is the easiest to understand. My resistance to embracing the explicit API here is partially due to the fact that we see so many apps await futures sequentially in main() when they can be parallelized. But this can probably be solved with documentation etc rather than withdrawing control.

That's interesting re `customer: quake`! I'm not aware of any google3 customers who do any kind of rendering before runApp. My perspective is skewed towards the handful of apps that I'm familiar with who basically load some data and initialize some services in main() and then render a conventional UI in runApp().

Reply:
I confirmed with `customer: quake` that they do indeed want to render non-text content prior to loading the fonts. How does that change your view?